workspaces/adventure-pack/goodies/typescript/Iterator.prototype.flatMap/index.test.ts

```diff
@@ -0,0 +1,184 @@
+    const generator = function* (
+      p: number,
+    ): Generator<number | boolean, void, void> {
+      yield p;
+      yield p % 2 === 0;
+    };
+
+    const flatMapResult = set.values().flatMap(generator);
+
+    expect([...flatMapResult]).toStrictEqual([
+      2,
+      true,
+      -4,
+      true,
+      7,
+      false,
+      9,
+      false,
+      -10,
+      true,
+    ]);
+  });
+
+  it("can map a Generator object", () => {
+    const factory = function* (): Generator<string, void, void> {
+      yield "o";
+      yield "t";
+      yield "t";
+      yield "e";
+      yield "r";
+    };
+
+    const generator = function* (c: string): Generator<string, void, void> {
+      yield c;
+      yield "!";
+    };
+
+    const flatMapResult = factory().flatMap(generator);
+
+    expect([...flatMapResult]).toStrictEqual([
+      "o",
+      "!",
+      "t",
+      "!",
+      "t",
+      "!",
+      "e",
+      "!",
+      "r",
+      "!",
+    ]);
+  });
+
+  it("returns an empty iterator when called on an empty iterator", () => {
+    const generator = function* <T>(p: T): Generator<T, void, void> {
+      yield p;
+    };
+
+    expect([].values().flatMap(generator).next().done).toBe(true);
+    expect(new Set().values().flatMap(generator).next().done).toBe(true);
+    expect(new Map().values().flatMap(generator).next().done).toBe(true);
+    expect((function* () {})().flatMap(generator).next().done).toBe(true);
+  });
+
+  it("callback returns an iterator but not an iterable", () => {
+    const iterator = [1, 2, 3].values();
+    const callback = <T>(x: T): Iterator<T> => {
+      let count = 0;
+      return Object.assign(Object.create(iteratorPrototype), {
+        next: () => {
+          if (count === 0) {
+            count++;
+            return { value: x };
+          }
+          return { done: true };
+        },
+      });
+    };
+
+    const flatMapResult = iterator.flatMap(callback);
+    expect([...flatMapResult]).toStrictEqual([1, 2, 3]);
+  });
+
+  it("callback returns an iterable but not an iterator", () => {
+    const iterator = [1, 2, 3].values();
+    const callback = <T>(x: T): Iterable<T> => ({
+      [Symbol.iterator]: function* () {
+        yield x;
+      },
+    });
+    const flatMapResult = iterator.flatMap(callback);
+    expect([...flatMapResult]).toStrictEqual([1, 2, 3]);
+  });
+
+  it("callback returns both an iterable and an iterator", () => {
+    const iterator = [1, 2, 3].values();
+    const callback = function* <T>(x: T) {
+      yield x;
+    };
+    const flatMapResult = iterator.flatMap(callback);
+    expect([...flatMapResult]).toStrictEqual([1, 2, 3]);
+  });
+
+  it("throws a TypeError when the callback does not return an iterator or iterable", () => {
+    const iterator = [1, 2, 3].values();
+    const invalidCallback = <T>(x: T): T => x;
+    // @ts-expect-error Incorrect callback return type
+    expect(() => iterator.flatMap(invalidCallback).next()).toThrow(TypeError);
```

Comment:
When using Jest's `toThrow`, I highly recommend isolating the part that is meant to throw. For example, do we have a throw when invoking `.flatMap` or when invoking `.next`? I am guessing it will throw on `.next`, right? If so:

```ts
const flatMapResult = iterator.flatMap(invalidCallback);
expect(() => flatMapResult.next()).toThrow(TypeError);
```